File: 2021/day2.cpp
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include "util.h"
using std::vector; using std::cout; using std::endl;
using std::string; using std::stoi;

vector<string> lines;

void part1() {
  int x = 0;
  int y = 0;
  for (int i = 0; i < lines.size(); i++) {
    vector<string> splitted = split(lines[i], ' ');
    string dir = splitted[0];
    int n = stoi(splitted[1]);
    if (dir == "forward") x += n;
    else if (dir == "up") y -= n;
    else if (dir == "down") y += n;
  }
  cout << x*y << endl;
}

void part2() {
  int x = 0;
  int y = 0;
  int a = 0;
  for (int i = 0; i < lines.size(); i++) {
    vector<string> splitted = split(lines[i], ' ');
    string dir = splitted[0];
    int n = stoi(splitted[1]);
    if (dir == "forward") {
      x += n;
      y += a*n;
    }
    else if (dir == "up") a -= n;
    else if (dir == "down") a += n;
  }
  cout << x*y << endl;
}

int main() {
  std::fstream data_file ("data/data2.txt", std::ios::in);
  string line;
  while (getline(data_file, line)) {
    lines.push_back(line);
  }
  data_file.close();

  part1();
  part2();
}

File: 2022/day2.py
p1 = 0
p2 = 0
with open("data/data2", "r") as f:
  for l in f:
    [a,b] = l.split(" ")
    a = a.strip()
    b = b.strip()

    if b == "X":
      p1 += 1
      if a == "A":
        p1 += 3
      elif a == "C":
        p1 += 6
    elif b == "Y":
      p1 += 2
      if a == "B":
        p1 += 3
      elif a == "A":
        p1 += 6
    elif b == "Z":
      p1 += 3
      if a == "C":
        p1 += 3
      elif a == "B":
        p1 += 6

    if b == "X":
      if a == "A":
        p2 += 3
      elif a == "B":
        p2 += 1
      else:
        p2 += 2
    elif b == "Y":
      p2 += 3
      if a == "A":
        p2 += 1
      elif a == "B":
        p2 += 2
      else:
        p2 += 3
    elif b == "Z":
      p2 += 6
      if a == "A":
        p2 += 2
      elif a == "B":
        p2 += 3
      else:
        p2 += 1

print(p1)
print(p2)